File: bench/erlang/pfib.erl
#!/usr/bin/env escript
%%! -mode(compile)
%% Parallel fib: compute fib(N) in 100 independent processes at once and sum the
%% results. The scheduler runs them across all cores. Checksum = 100 * fib(N).
main(_) ->
    N = bench_n(28), Tasks = 100,
    Parent = self(),
    lists:foreach(fun(_) -> spawn(fun() -> Parent ! {r, fib(N)} end) end, lists:seq(1, Tasks)),
    Total = collect(Tasks, 0),
    io:format("~w~n", [Total]).
bench_n(D) -> case os:getenv("BENCH_N") of false -> D; V -> list_to_integer(V) end.
collect(0, Acc) -> Acc;
collect(K, Acc) -> receive {r, V} -> collect(K-1, Acc+V) end.
fib(N) when N < 2 -> N;
fib(N) -> fib(N-1) + fib(N-2).

File: bench/go/pfib.go
// Parallel fib: compute fib(N) in 100 goroutines at once and sum the results; the
// scheduler spreads them across every core. Checksum = 100 * fib(N).
package main

import "fmt"

func fib(n int) int {
	if n < 2 {
		return n
	}
	return fib(n-1) + fib(n-2)
}

func main() {
	n := benchN(31)
	const tasks = 100
	results := make(chan int, tasks)
	for i := 0; i < tasks; i++ {
		go func() { results <- fib(n) }()
	}
	total := 0
	for i := 0; i < tasks; i++ {
		total += <-results
	}
	fmt.Println(total)
}
